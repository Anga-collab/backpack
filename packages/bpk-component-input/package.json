{
  "name": "bpk-component-input",
  "version": "6.0.39",
  "description": "Backpack input component.",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "git@github.com:Skyscanner/backpack.git"
  },
  "author": "Backpack Design System <backpack@skyscanner.net>",
  "main": "index.js",
  "publishConfig": {
    "registry": "https://registry.npmjs.org/"
  },
  "gitHead": "4e4ec9add28e1db54d6a38a10c331bbd163f2926",
  "dependencies": {
    "bpk-component-icon": "^9.1.13",
    "bpk-mixins": "^25.2.0",
    "bpk-react-utils": "^4.0.2",
    "prop-types": "^15.7.2"
  },
  "peerDependencies": {
    "react": "^16.3.0"
  },
  "devDependencies": {
    "bpk-component-banner-alert": "^6.0.12",
    "bpk-component-label": "^5.0.25",
    "bpk-storybook-utils": "^1.0.25"
  }
}
