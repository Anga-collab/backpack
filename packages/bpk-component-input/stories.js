import {
  TextExample,
  PlaceholderExample,
  ValidExample,
  InvalidExample,
  DisabledExample,
  ClearableExample,
  EmailInputExample,
  NumberInputExample,
  PasswordInputExample,
  TelephoneInputExample,
  LargeInputExample,
  DockedExample,
  ManuallyDockedExample,
  MixedExample,
} from './examples';

export default {
  title: 'bpk-component-input',
};

export const TextValue = TextExample;

TextValue.story = {
  name: 'Text value',
};

export const Placeholder = PlaceholderExample;
export const Valid = ValidExample;
export const Invalid = InvalidExample;
export const Disabled = DisabledExample;
export const Clearable = ClearableExample;
export const Email = EmailInputExample;
export const Number = NumberInputExample;
export const Password = PasswordInputExample;
export const Telephone = TelephoneInputExample;
export const Large = LargeInputExample;
export const Docked = DockedExample;
export const ManuallyDocked = ManuallyDockedExample;

ManuallyDocked.story = {
  name: 'Manually docked',
};

export const VisualTest = MixedExample;

VisualTest.story = {
  name: 'Visual test',
};
