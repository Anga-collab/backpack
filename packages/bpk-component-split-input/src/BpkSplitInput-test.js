/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/* @flow strict */

import React from 'react';
import { shallow, mount } from 'enzyme';
import { INPUT_TYPES } from 'bpk-component-input';

import BpkSplitInput from './BpkSplitInput';

const defaultProps = {
  name: 'otpInput',
  id: 'otpInput',
  label: 'otp input',
  onChange: () => {},
  onSubmit: () => {},
};

const typeInInput = (input, element) => {
  const inputChars = input.split('');
  for (let i = 0; i < inputChars.length; i += 1) {
    const event = { target: { value: inputChars[i] } };
    element.find('input').at(i).simulate('change', event);
  }
};

const pasteInInput = (input, element) => {
  const pasteEvent = {
    clipboardData: { getData: jest.fn().mockReturnValueOnce(input) },
  };

  element.find('input').at(0).simulate('paste', pasteEvent);
};

describe('BpkSplitInput', () => {
  const mockOnChange = jest.fn();
  const mockOnSubmit = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render without crashing', () => {
    const wrapper = shallow(<BpkSplitInput {...defaultProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with inputLength param', () => {
    const wrapper = shallow(
      <BpkSplitInput {...defaultProps} inputLength={6} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with large set as false', () => {
    const wrapper = shallow(<BpkSplitInput {...defaultProps} large={false} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with placeholder', () => {
    const wrapper = shallow(
      <BpkSplitInput {...defaultProps} placeholder="x" />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should call onChange callback function on every valid numeric input', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );

    expect(bpkSplitInput.state().inputValue).toEqual([]);

    typeInInput('9234', bpkSplitInput);

    expect(bpkSplitInput.state().inputValue).toEqual(['9', '2', '3', '4']);
    expect(mockOnChange).toHaveBeenCalledTimes(4);
    expect(mockOnChange).toHaveBeenLastCalledWith('9234');
  });

  it('should call onChange callback function on every valid char input', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        type={INPUT_TYPES.text}
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );

    expect(bpkSplitInput.state().inputValue).toEqual([]);

    typeInInput('xyz6', bpkSplitInput);

    expect(bpkSplitInput.state().inputValue).toEqual(['x', 'y', 'z', '6']);
    expect(mockOnChange).toHaveBeenCalledTimes(4);
    expect(mockOnChange).toHaveBeenCalledWith('xyz6');
  });

  it('should call onChange callback function only on valid numeric input', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );

    expect(bpkSplitInput.state().inputValue).toEqual([]);

    typeInInput('92x4', bpkSplitInput);

    expect(bpkSplitInput.state().inputValue).toEqual(['9', '2', '4']);
    expect(mockOnChange).toHaveBeenCalledTimes(3);
    expect(mockOnChange).toHaveBeenLastCalledWith('924');
  });

  it('should not call onChange callback function on invalid input', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );

    expect(bpkSplitInput.state().inputValue).toEqual([]);

    typeInInput('$', bpkSplitInput);

    expect(bpkSplitInput.state().inputValue).toEqual([]);

    expect(mockOnChange).not.toHaveBeenCalled();
  });

  it('should update the focusedInput state on valid input', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );

    expect(bpkSplitInput.state().focusedInput).toEqual(0);

    typeInInput('9', bpkSplitInput);

    expect(bpkSplitInput.state().focusedInput).toEqual(1);
  });

  it('should update the focusedInput state on focus to an input', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );

    expect(bpkSplitInput.state().focusedInput).toEqual(0);

    const event = { target: { select: jest.fn() } };
    bpkSplitInput.find('input').at(2).prop('onFocus')(event);

    expect(bpkSplitInput.state().focusedInput).toEqual(2);
  });

  it('should update the focusedInput state and the inputValue on pressing Backspace', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );

    expect(bpkSplitInput.state().focusedInput).toEqual(0);

    typeInInput('9234', bpkSplitInput);

    expect(bpkSplitInput.state().focusedInput).toEqual(3);

    bpkSplitInput.find('input').at(3).simulate('keydown', { key: 'Backspace' });

    expect(bpkSplitInput.state().focusedInput).toEqual(2);
    expect(mockOnChange).toHaveBeenLastCalledWith('923');
  });

  it('should update the input value but not the focusedInput state on pressing Delete', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );

    expect(bpkSplitInput.state().focusedInput).toEqual(0);

    typeInInput('9234', bpkSplitInput);

    expect(bpkSplitInput.state().focusedInput).toEqual(3);

    bpkSplitInput.find('input').at(3).simulate('keydown', { key: 'Delete' });

    expect(bpkSplitInput.state().focusedInput).toEqual(3);
    expect(mockOnChange).toHaveBeenLastCalledWith('923');
  });

  it('should update the focusedInput state without updating the value on pressing Left arrow key', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );

    expect(bpkSplitInput.state().focusedInput).toEqual(0);

    typeInInput('9234', bpkSplitInput);

    expect(bpkSplitInput.state().focusedInput).toEqual(3);

    bpkSplitInput.find('input').at(3).simulate('keydown', { key: 'Left' });

    expect(bpkSplitInput.state().focusedInput).toEqual(2);
    expect(mockOnChange).toHaveBeenLastCalledWith('9234');
  });

  it('should not update the focusedInput state or the value on pressing Left arrow key when on the left most input', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );

    expect(bpkSplitInput.state().focusedInput).toEqual(0);

    bpkSplitInput.find('input').at(0).simulate('keydown', { key: 'Left' });

    expect(bpkSplitInput.state().focusedInput).toEqual(0);
    expect(mockOnChange).not.toHaveBeenCalled();
  });

  it('should update the focusedInput state without updating the value on pressing Right arrow key', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );

    expect(bpkSplitInput.state().focusedInput).toEqual(0);

    typeInInput('9234', bpkSplitInput);

    expect(bpkSplitInput.state().focusedInput).toEqual(3);

    bpkSplitInput.find('input').at(2).simulate('keydown', { key: 'Right' });

    expect(bpkSplitInput.state().focusedInput).toEqual(3);
    expect(mockOnChange).toHaveBeenLastCalledWith('9234');
  });

  it('should not update the focusedInput state or the value on pressing Right arrow key when on the Right most input', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );

    expect(bpkSplitInput.state().focusedInput).toEqual(0);

    typeInInput('9234', bpkSplitInput);

    expect(bpkSplitInput.state().focusedInput).toEqual(3);

    bpkSplitInput.find('input').at(3).simulate('keydown', { key: 'Right' });

    expect(bpkSplitInput.state().focusedInput).toEqual(3);
    expect(mockOnChange).toHaveBeenLastCalledWith('9234');
  });

  it('should not update the focusedInput state or the value on pressing space bar', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );

    expect(bpkSplitInput.state().focusedInput).toEqual(0);

    bpkSplitInput.find('input').at(0).simulate('keydown', { key: 'Spacebar' });

    expect(bpkSplitInput.state().focusedInput).toEqual(0);
    expect(mockOnChange).not.toHaveBeenCalled();
  });

  it('should call onSubmit callback when the input is valid and Enter key is pressed', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        inputLength={6}
        onSubmit={mockOnSubmit}
      />,
    );

    typeInInput('923467', bpkSplitInput);

    bpkSplitInput.find('input').at(0).simulate('keydown', { keyCode: 13 });

    expect(mockOnSubmit).toHaveBeenCalledWith('923467');
  });

  it('should spread the pasted text over multiple input fields on valid input', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );

    pasteInInput('1234', bpkSplitInput);

    expect(bpkSplitInput.state().focusedInput).toEqual(3);
    expect(mockOnChange).toHaveBeenCalledWith('1234');
  });

  it('should spread the valid pasted input over multiple input fields and focus the first invalid input', async () => {
    const bpkSplitInput = mount(
      <BpkSplitInput
        name="otpInput"
        id="otpInput"
        label="otp input"
        onChange={mockOnChange}
        onSubmit={() => {}}
      />,
    );
    pasteInInput('12x4', bpkSplitInput);

    expect(bpkSplitInput.state().focusedInput).toEqual(2);
    expect(mockOnChange).toHaveBeenCalledWith('124');
  });
});
