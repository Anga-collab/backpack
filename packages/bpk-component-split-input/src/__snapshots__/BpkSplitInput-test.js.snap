// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkSplitInput should render correctly with inputLength param 1`] = `
<div
  className="BpkSplitInput"
>
  <BpkInputField
    focus={true}
    id="otpInput-0"
    index={0}
    key="0"
    label="otp input"
    large={true}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=""
    type="number"
    value=""
  />
  <BpkInputField
    focus={false}
    id="otpInput-1"
    index={1}
    key="1"
    label="otp input"
    large={true}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=""
    type="number"
    value=""
  />
  <BpkInputField
    focus={false}
    id="otpInput-2"
    index={2}
    key="2"
    label="otp input"
    large={true}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=""
    type="number"
    value=""
  />
  <BpkInputField
    focus={false}
    id="otpInput-3"
    index={3}
    key="3"
    label="otp input"
    large={true}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=""
    type="number"
    value=""
  />
  <BpkInputField
    focus={false}
    id="otpInput-4"
    index={4}
    key="4"
    label="otp input"
    large={true}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=""
    type="number"
    value=""
  />
  <BpkInputField
    focus={false}
    id="otpInput-5"
    index={5}
    key="5"
    label="otp input"
    large={true}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=""
    type="number"
    value=""
  />
</div>
`;

exports[`BpkSplitInput should render correctly with large set as false 1`] = `
<div
  className="BpkSplitInput"
>
  <BpkInputField
    focus={true}
    id="otpInput-0"
    index={0}
    key="0"
    label="otp input"
    large={false}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=""
    type="number"
    value=""
  />
  <BpkInputField
    focus={false}
    id="otpInput-1"
    index={1}
    key="1"
    label="otp input"
    large={false}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=""
    type="number"
    value=""
  />
  <BpkInputField
    focus={false}
    id="otpInput-2"
    index={2}
    key="2"
    label="otp input"
    large={false}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=""
    type="number"
    value=""
  />
  <BpkInputField
    focus={false}
    id="otpInput-3"
    index={3}
    key="3"
    label="otp input"
    large={false}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=""
    type="number"
    value=""
  />
</div>
`;

exports[`BpkSplitInput should render correctly with placeholder 1`] = `
<div
  className="BpkSplitInput"
>
  <BpkInputField
    focus={true}
    id="otpInput-0"
    index={0}
    key="0"
    label="otp input"
    large={true}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder="x"
    type="number"
    value=""
  />
  <BpkInputField
    focus={false}
    id="otpInput-1"
    index={1}
    key="1"
    label="otp input"
    large={true}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder="x"
    type="number"
    value=""
  />
  <BpkInputField
    focus={false}
    id="otpInput-2"
    index={2}
    key="2"
    label="otp input"
    large={true}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder="x"
    type="number"
    value=""
  />
  <BpkInputField
    focus={false}
    id="otpInput-3"
    index={3}
    key="3"
    label="otp input"
    large={true}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder="x"
    type="number"
    value=""
  />
</div>
`;

exports[`BpkSplitInput should render without crashing 1`] = `
<div
  className="BpkSplitInput"
>
  <BpkInputField
    focus={true}
    id="otpInput-0"
    index={0}
    key="0"
    label="otp input"
    large={true}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=""
    type="number"
    value=""
  />
  <BpkInputField
    focus={false}
    id="otpInput-1"
    index={1}
    key="1"
    label="otp input"
    large={true}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=""
    type="number"
    value=""
  />
  <BpkInputField
    focus={false}
    id="otpInput-2"
    index={2}
    key="2"
    label="otp input"
    large={true}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=""
    type="number"
    value=""
  />
  <BpkInputField
    focus={false}
    id="otpInput-3"
    index={3}
    key="3"
    label="otp input"
    large={true}
    name="otpInput"
    onChange={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=""
    type="number"
    value=""
  />
</div>
`;
